### PR TITLE
fix possible deadlock in when syncing segments to datanode(#25196)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #25196

### CODE REVIEW

File: internal/datanode/data_node.go [hunk 1]
```diff
@@ -948,8 +948,15 @@ func (node *DataNode) SyncSegments(ctx context.Context, req *datapb.SyncSegments
...
-	// block all flow graph so it's safe to remove segment
-	ds.fg.Blockall()
+	// block all flow graph, so it's safe to remove segment
+	if !ds.fg.TryBlockAll(ctx) {
+		log.Ctx(ctx).Error("Cannot block flowGraph for sync segments, "+
+			"release generated segment meta and stats log", zap.Int64("planID", req.GetPlanID()),
+			zap.Int64("target segmentID", req.GetCompactedTo()))
+		targetSeg.historyStats = nil
+		status.Reason = "failed to block flowgraph"
+		return status, errors.New("failed to block flowgraph")
```
Comment (jaime0815) [inline internal/datanode/data_node.go:958]: compactionExecutor also should be done with a failure status.

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/util/lock/lock_util.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/milvus-io/milvus/internal/util/logutil"
+	"github.com/milvus-io/milvus/internal/util/retry"
+	"go.uber.org/zap"
+)
+
+func TryLock(ctx context.Context, source string, mutex *sync.RWMutex,
+	interval time.Duration, retryCount uint, printDuration bool) bool {
+	log := logutil.Logger(ctx).With(zap.String("source", source))
+	beforeTryLock := time.Now()
+	err := retry.Do(ctx, func() error {
```
Comment (jaime0815) [inline internal/util/lock/lock_util.go:34]: Try lock should fail while ctx  receive done

Comment (MrPresent-Han) [reply internal/util/lock/lock_util.go:34]: retry can cover context cancel case

File: internal/util/flowgraph/flow_graph.go [hunk 3]
```diff
@@ -20,16 +20,27 @@ import (
...
 // Flow Graph is no longer a graph rather than a simple pipeline, this simplified our code and increase recovery speed - xiaofan.
 
 // TimeTickedFlowGraph flowgraph with input from tt msg stream
 type TimeTickedFlowGraph struct {
-	nodeCtx   map[NodeName]*nodeCtx
-	stopOnce  sync.Once
-	startOnce sync.Once
-	closeWg   *sync.WaitGroup
+	nodeCtx      map[NodeName]*nodeCtx
+	nodeSequence []string
```
Comment (jaime0815) [inline internal/util/flowgraph/flow_graph.go:40]: Please check other flowgraph operations by nodeCtx should follow the order of the node in nodeSequence.

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by MrPresent-Han** [comment]
rerun ut

**Comment by bigsheeper** [comment]
LGTM but why there is a deadlock possibility in the original `BlockAll` function?

**Comment by MrPresent-Han** [comment]
> LGTM but why there is a deadlock possibility in the original `BlockAll` function?

original block all used nodeCtx map, which provides unstable keys order for every iteration

**Review by congqixia — APPROVED** [review]
/lgtm

**Comment by congqixia** [comment]
@MrPresent-Han please cherry-pick to master, thx

**Pull request merged**